lib/integrations/CalDav/CalDavCalendarAdapter.ts

```diff
@@ -0,0 +1,301 @@
+import { IntegrationCalendar, CalendarApiAdapter, CalendarEvent } from "../../calendarClient";
+import { symmetricDecrypt } from "@lib/crypto";
+import {
+  createAccount,
+  fetchCalendars,
+  fetchCalendarObjects,
+  getBasicAuthHeaders,
+  createCalendarObject,
+  updateCalendarObject,
+  deleteCalendarObject,
+} from "tsdav";
+import { Credential } from "@prisma/client";
+import ICAL from "ical.js";
+import { createEvent, DurationObject, Attendee, Person } from "ics";
+import dayjs from "dayjs";
+import { v4 as uuidv4 } from "uuid";
+
+type EventBusyDate = Record<"start" | "end", Date>;
+
+export class CalDavCalendar implements CalendarApiAdapter {
+  private url: string;
+  private credentials: Record<string, string>;
+  private headers: Record<string, string>;
+  private readonly integrationName: string = "caldav_calendar";
+
+  constructor(credential: Credential) {
+    const unencryptedCrendential = JSON.parse(
+      symmetricDecrypt(credential.key, process.env.CALENDSO_ENCRYPTION_KEY)
+    );
+    const username = unencryptedCrendential.username;
+    const url = unencryptedCrendential.url;
+    const password = unencryptedCrendential.password;
+
+    this.url = url;
+
+    this.credentials = {
+      username,
+      password,
+    };
+
+    this.headers = getBasicAuthHeaders({
+      username,
+      password,
+    });
+  }
+
+  convertDate(date: string): number[] {
+    return dayjs(date)
+      .utc()
+      .toArray()
+      .slice(0, 6)
+      .map((v, i) => (i === 1 ? v + 1 : v));
+  }
+
+  getDuration(start: string, end: string): DurationObject {
+    return {
+      minutes: dayjs(end).diff(dayjs(start), "minute"),
+    };
+  }
+
+  getAttendees(attendees: Person[]): Attendee[] {
+    return attendees.map(({ email, name }) => ({ name, email, partstat: "NEEDS-ACTION" }));
+  }
+
+  stripHtml(str: string) {
+    return str
+      .replace(/<br\s?\/>/g, "\n")
+      .replace(/<a.+?href="mailto:.+?>(.+?)<\/a>/g, "$1")
+      .replace(/<strong>|<\/strong>/g, "")
+      .replace(/<br\s{0,}}]\/>/g, "")
+      .replace(/<p.+?>|<\/p>/g, "");
+  }
```

Comment:
There's a similar exported function (`stripHtml`) in `lib/emails/helpers.ts`. You could consider moving the helpers.ts to the parent folder (since it can be seen in a more general way), maybe adapt it somehow and reuse it in your code here. "My" function replaces the HTML codes in a more general way, but has explicit regular expressions for a-links and a-mailtos as well:
![image](https://user-images.githubusercontent.com/14024119/128776329-8f171dc3-6568-49fd-9879-eac7b04868a0.png)